keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,737 @@
+finding the Primary Node IP is still the same as it used to be; we
+will just adjust `Node.Status.Addresses` to ensure that someone
+following the traditional rule will find the IP we want them to.
+
+There is now also a "Secondary Node IP", which is the first
+`InternalIP` address of the opposite IP family from the Primary Node
+IP, or the first such `ExternalIP` if there is no such `InternalIP`.
+
+#### Updated Use of `Node.Status.Addresses`
+
+When setting Pod status, kubelet will continue to use the Primary Node
+IP for `Pod.Status.HostIP`.
+
+In clusters with the `IPv6DualStack` feature gate enabled (or
+post-dual-stack GA), host-network Pods will now get both the Primary
+Node IP and the Secondary Node IP as `Pod.Status.PodIPs` (allowing
+host-network Pods to be endpoints of both IPv4 and IPv6 Services).
+
+Some methods in `utilnode` such as `GetNodeHostIP` and
+`GetPreferredNodeAddress` may be updated to have dual-stack versions
+that return multiple IPs. The existing versions are still useful on
+their own though.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+TODO
+
+## Design Details
+
+### Implementation
+
+As before, when using an external cloud provider, the external
+provider deals with setting `Node.Status.Addresses`, and when using a
+legacy cloud provider or bare metal, kubelet deals with setting it.
+
+#### With a Cloud Provider (External or Legacy)
+
+When using an external cloud provider, kubelet will pass the value of
+`--node-ips` (explicit or defaulted) to the provider via a Node
+annotation. (If the provider finds that kubelet has not set the
+`--node-ips` annotation, then it should assume it's running against an
+old kubelet, and behave as it would have in previous releases.)
+
+Kubelet or the external cloud provider will now get the provisional
+list of IPv4 and IPv6 node addresses from the cloud provider-specific
+code (which will hopefully have been updated to return IPv6 node IPs)
+and then modify the list of node addresses as follows. ("Kubelet"
+below really means "Either kubelet or the external cloud provider".)
+
+  1. When using a legacy cloud provider, if the deprecated `--node-ip`
+     argument was passed, then kubelet will first do the filtering and
+     sorting associated with that.
+
+  2. If `--node-ips` has a single element, then all addresses of the
+     non-matching family will be removed from the list.
+
+  3. If there are no longer any `InternalIP` or `ExternalIP` elements
+     in the list, then the provided `--node-ips` is invalid, and
+     kubelet will error out.
+
+  4. If either element of `--node-ips` is a specific IP address, and
```

Comment:
It could... After feedback on the original much-broader proposal I was trying to keep this one much more limited in scope though. And I haven't ever seen any request for that feature (though I can see how it would be useful on hosts attached to multiple networks)

The existing `--node-ip` flag uses `0.0.0.0` to mean "any IPv4" and `::` to mean "any IPv6", but I thought it would be nicer to use `ipv4` and `ipv6` in `--node-ips` instead... but in fact, the code just converts them to `0.0.0.0` and `::` internally (https://github.com/danwinship/kubernetes/blob/1b229b63/pkg/util/node/node.go#L395). So maybe?

FWIW if we only accepted IPs now, we could compatibly extend it to allow CIDRs as well later.

We'd have to rethink the rules about optional vs mandatory; if `--node-ips 10.0.0.1/32` means "has to have the IP 10.0.0.1" then `--node-ips 0.0.0.0/0,::/0` seems like it should mean "has to have some IPv4 address and some IPv6 address", but currently `--node-ips ipv4,ipv6` is spec'ed to mean "_may_ have some IPv4 address and some IPv6 address". I think the default behavior should be "either single-stack IPv4 or single-stack IPv6 or dual-stack depending on what's available", and if it is there has to be some way to say that.